Add To/FromField instances for UTCTime.

dmjio opened the pull request:
- [x] Support fractional seconds in `ToField`
- [x] Support timezone offsets

dmjio committed:
Add To/FromField instances for UTCTime.

dmjio commented on src/Database/Oracle/Simple/FromField.hs:
```diff
@@ -62,14 +63,17 @@ instance (FromField a) => FromField (Maybe a) where
+          tz = utc { timeZoneMinutes = fromIntegral $ (60 * tzHourOffset) + tzMinuteOffset }
+          local = LocalTime d tod
+          d = fromGregorian (fromIntegral year) (fromIntegral month) (fromIntegral day)
+          tod = TimeOfDay (fromIntegral hour) (fromIntegral minute) (fromIntegral second)
```
Need to incorporate `fsecond` here

apkhandkar commented on src/Database/Oracle/Simple/ToField.hs:
```diff
@@ -27,3 +30,25 @@ instance ToField Int where
   toField = pure . AsTimestamp
+
+instance ToField UTCTime where
+  toField utcTime = pure (AsTimestamp dpiTs)
```
This could also be exported as a "smart" constructor for DPITimestamp
```haskell
mkDPITimestamp :: UTCTime -> DPITimestamp
mkDPITimestamp utcTime = ...

instance ToField UTCTime where
  toField = pure . AsTimestamp . mkDPITimestamp
```

dmjio replied:
That's a great idea, yea we probably shouldn't export the type at all, just keep it internal. The conversion functions can be used in `ToField` / `FromField` and then we can those functions as well for roundtrip property-based tests.

apkhandkar reviewed (commented):
If we have ToField/FromField instances for UTCTime, do we need to have ToField/FromField instances for DPITimestamp, or expose the type at all? It can be treated as an "internal" type, how we treat DPIBytes. We can then have a `ReadDPIBuffer UTCTime` like so:
```haskell
instance FromField UTCTime where
  fromField = fieldParser getUTCTime

getUTCTime :: ReadDPIBuffer UTCTime
getUTCTime = convert <=< peek <=< dpiData_getTimestamp
  where
    convert DPITimestamp {..} = ...
```
Perhaps expose the `convert` function as `dpiTimestampToUTCTime`.

dmjio commented:
Consider how `odpic-raw` did this.

dmjio committed:
Support timezone offsets.

Converts all DPITimeStamp into UTC first, before converting to UTCTime.

dmjio committed:
Use describe block for DpiTimestamp

liminalisht commented on src/Database/Oracle/Simple/Internal.hs:
```diff
@@ -541,9 +543,34 @@ data DPITimestamp = DPITimestamp
+    minute         <- choose (1, 59)
+    second         <- choose (1, 59)
+    fsecond        <- choose (0, 100000)
+    tzHourOffset   <- choose (-11, 11)
```
should be `choose (-11, 11)`

dmjio replied:
derp, updated !

dmjio committed:
Use +/- 14 hour offsets

liminalisht commented:
I added this test to your branch
```
      it "Should roundtrip UTCTime through DPITimestamp" $ \conn -> do
        property $ \utc -> do
          let dpi = (utcTimeToDPITimestamp utc)
          diffUTCTime (dpiTimeStampToUTCTime dpi) utc `shouldBe` 0
```
to see how roundtripping starting w/ a UTCTime would work:
```
  1) SELECT tests Should roundtrip UTCTime through DPITimestamp
       Falsifiable (after 1 test and 9 shrinks):
         1858-11-17 04:16:00 UTC
       expected: 0s
        but got: -15360s
```

dmjio commented:
So we need to convert the `DPITimestamp` to UTC first, which the `ToField` / `FromField` will always ensure that happens. Your test should pass with this update.

If you'd like to add this test in a separate MR, that'd be great too.

```haskell
      it "Should roundtrip UTCTime through DPITimestamp" $ \conn -> do                                                                                                                                                                      
        property $ \utc -> do                                                                                                                                                                                                               
          let dpi = (utcTimeToDPITimestamp utc)                                                                                                                                                                                             
          diffUTCTime (dpiTimeStampToUTCTime (dpiTimeStampToUTCDPITimeStamp dpi)) utc `shouldBe` 0     
```

dmjio committed:
Hour should be up to 23

dmjio committed:
Offset should affect day, month, year potentially.

dmjio committed:
Add documentation.

dmjio committed:
Move dpiTimeStampToUTCDPITimeStamp to Internal
  - Apply dpiTimeStampToUTCDPITimeStamp in ToField
  - Use utc in dpiTimeStampToUTCTime

dmjio committed:
Add idempotency tests to DpiTimestamp

apkhandkar commented on src/Database/Oracle/Simple/Internal.hs:
```diff
@@ -541,9 +543,60 @@ data DPITimestamp = DPITimestamp
+      , day = fromIntegral day'
+      , hour = hours
+      , minute = minutes
+      }
```
The `Int8`/`Word8` arithmetic here is causing (what I presume are) overflows.
```haskell
> :t istDpi -- DPITimestamp with offset of +05:30
istDpi :: DPITimestamp
> :t hour istDpi
hour istDpi :: GHC.Word.Word8
> hour istDpi
11
> (hour istDpi * 60) -- as seen in the fn
148 -- whoops! should be 660!
```
In turn, the converted UTC DPI timestamp (and by extension, UTC time) is incorrect:
```haskell
> istDpi
DPITimestamp {year = 2023, month = 9, day = 22, hour = 11, minute = 50, second = 2, fsecond = 770910000, tzHourOffset = 5, tzMinuteOffset = 30}
> dpiTimeStampToUTCTime istDpi 
2023-09-22 02:04:02.00077091 UTC -- should be 6AM UTC due to the +05:30 offset!
```

apkhandkar replied:
This test highlights the issue:
```haskell
      it "Should select timestamp from Oracle" $ \conn -> do
        utcTime <- getCurrentTime
        [getOnly -> gotUtcTime] <- query_ @(Only UTCTime) conn "select current_timestamp from dual"
        gotUtcTime `shouldBe` utcTime
```
```
Failures:

  test/Main.hs:29:17: 
  1) SELECT tests Should select timestamp from Oracle      
        expected: 2023-09-22 06:58:03.401534318 UTC
        but got: 2023-09-22 02:42:03.000401937 UTC

  To rerun use: --match "/SELECT tests/Should select timestamp from Oracle/"
```

apkhandkar commented on test/Main.hs:
```diff
@@ -24,6 +28,19 @@ spec = do
+        property $ \dpi -> do
+          dpiTimeStampToUTCDPITimeStamp (dpiTimeStampToUTCDPITimeStamp dpi)
+            `shouldBe` dpiTimeStampToUTCDPITimeStamp dpi
+
```
These tests are not picking up on the DPITimestamp conversion issue noted above

apkhandkar reviewed (changes requested):
A rough fix that addresses the overflow issue:
```haskell
-- | Converts a DPITimestamp into the UTCTime zone by applying the offsets
-- to the year, month, day, hour, minutes and seconds
dpiTimeStampToUTCDPITimeStamp :: DPITimestamp -> DPITimestamp
dpiTimeStampToUTCDPITimeStamp dpi@DPITimestamp{..} = utcDpi
  where
    offsetInMinutes = (fromIntegral tzHourOffset * 60) + fromIntegral tzMinuteOffset
    currentMinutes = (fromIntegral hour * 60) + fromIntegral minute
    (fromIntegral -> hours, fromIntegral -> minutes) = (currentMinutes - offsetInMinutes) `quotRem` 60
    gregorianDay = fromGregorian (fromIntegral year) (fromIntegral month) (fromIntegral day)
    updatedDay | fromIntegral currentMinutes + fromIntegral offsetInMinutes > 1440
               = addDays 1 gregorianDay
               | fromIntegral currentMinutes + fromIntegral offsetInMinutes < 0
               = addDays (-1) gregorianDay
               | otherwise = gregorianDay
    (year', month', day') = toGregorian updatedDay
    utcDpi
      = dpi
      { tzHourOffset = 0
      , tzMinuteOffset = 0
      , year = fromIntegral year'
      , month = fromIntegral month'
      , day = fromIntegral day'
      , hour = hours
      , minute = minutes
      }
```
We shouldn't expect the exact UTC timestamp, but I'd say we're close enough:
```
Failures:

  test/Main.hs:29:17: 
  1) SELECT tests Should select timestamp from Oracle
       expected: 2023-09-22 07:29:00.000698106 UTC
        but got: 2023-09-22 07:29:00.697772802 UTC

  To rerun use: --match "/SELECT tests/Should select timestamp from Oracle/"

```
This breaks the round trip tests, so the fix needs fixing...
```
  test/Main.hs:40:13: 
  2) DPITimeStamp tests Should roundtrip DPITimestamp through UTCTime
       Falsifiable (after 5 tests):
         DPITimestamp {year = 1798, month = 1, day = 12, hour = 23, minute = 48, second = 59, fsecond = 94601, tzHourOffset = -10, tzMinuteOffset = 42}
       expected: DPITimestamp {year = 1798, month = 1, day = 12, hour = 33, minute = 6, second = 59, fsecond = 94601, tzHourOffset = 0, tzMinuteOffset = 0}
        but got: DPITimestamp {year = 1798, month = 1, day = 13, hour = 9, minute = 6, second = 59, fsecond = 94601, tzHourOffset = 0, tzMinuteOffset = 0}

  To rerun use: --match "/DPITimeStamp tests/Should roundtrip DPITimestamp through UTCTime/"

  test/Main.hs:45:13: 
  3) DPITimeStamp tests Idempotency of dpiTimeStampToUTCDPITimeStamp 
       Falsifiable (after 5 tests):
         DPITimestamp {year = 1798, month = 1, day = 12, hour = 23, minute = 48, second = 59, fsecond = 94601, tzHourOffset = -10, tzMinuteOffset = 42}
       expected: DPITimestamp {year = 1798, month = 1, day = 12, hour = 33, minute = 6, second = 59, fsecond = 94601, tzHourOffset = 0, tzMinuteOffset = 0}
        but got: DPITimestamp {year = 1798, month = 1, day = 13, hour = 33, minute = 6, second = 59, fsecond = 94601, tzHourOffset = 0, tzMinuteOffset = 0}
```

dmjio committed:
Address issues with UTC conversion on DPITimestamp
  - Addresses overflow bug w/ fromIntegral conversion
  - Negates offset before adding to current minutes
  - mod by 1440 (minutes in a day) to address day overflow
  - Addressed bug where minutes and hour offsets did not have the same
  signum.
  - Added test for YYYY/MM/DD changes due to offset.

Co-authored-by: Arjun Khandkar <khandkararjun@gmail.com>

dmjio commented:
@apkhandkar and I have addressed year / month / day changes when dealing with timezone offsets for `DPITimestamp`. `UTCTime` round trip still seems to be suffering from 0 > ~800ms loss of time during conversion (specifically on the initial conversion to DPITimestamp). See [4a63b63](https://github.com/haskell-oracle/oracle-simple/pull/13/commits/4a63b63970c1e61ff117bbfa418e029895e44311) for more.

dmjio committed:
Add isomorphism tests for UTCTime -> DPITimestamp -> UTCTime, with the
caveat that only nanosecond granularity is used.

Co-authored-by: Arjun Khandkar <khandkararjun@gmail.com>

liminalisht commented on src/Database/Oracle/Simple/FromField.hs:
```diff
@@ -62,14 +65,19 @@ instance (FromField a) => FromField (Maybe a) where
+    local = LocalTime d tod
+    d = fromGregorian (fromIntegral year) (fromIntegral month) (fromIntegral day)
+    tod = TimeOfDay (fromIntegral hour) (fromIntegral minute) (fromIntegral second + picos)
+    picos = MkFixed (fromIntegral fsecond) :: Pico
```
`fsecond` in the struct expects nanos, not picos
https://odpi-c.readthedocs.io/en/latest/structs/dpiTimestamp.html

I'd convert from nanos to picos here.

liminalisht commented on src/Database/Oracle/Simple/ToField.hs:
```diff
@@ -27,3 +31,28 @@ instance ToField Int where
+      , hour           = fromIntegral todHour
+      , minute         = fromIntegral todMin
+      , second         = seconds
+      , fsecond        = truncate (fractionalSeconds * 1_000_000_000_000)
```
`fsecond` in the struct expects nanos, not picos
https://odpi-c.readthedocs.io/en/latest/structs/dpiTimestamp.html

I'd convert from picos to nanos here.

dmjio replied:
Dunzo

dmjio committed:
Properly convert Nanoseconds to Picoseconds and back.

Gives isomorphism property on UTCTime for nanosecond precision.

Co-authored-by: Abraham Howland <ahowland@gmail.com>
